File: XyzPython/XyzMessage.py
from XyzPython.XyzUtils import XyzUtils
import struct

class DataError(Exception):
    '''
    Exception raised when there is not enough
    data in the XyzMessage buffer to generate
    a requested data type.
    '''
    pass

class XyzMessage:
    '''
    A class for generating/unpacking valid Messages for the Xyz protocol.

    Parameters:
        data\t- The decrypted, inflated data of the packet (Bytes)
        msgid\t- The Message ID of the packet.
    '''
    def __init__(self, data: bytes, msgid: int=0):
        self.data = data
        self.msgid = msgid

    def get(self):
        '''
        Deflates and packs the packet data/id and returns bytes that can be sent
        as-is to the server/client.
        '''
        data = XyzUtils.deflate(self.data)
        return struct.pack("IB", len(data), self.msgid)+data

    def bools(self):
        '''
        Unpacks the data to a list of booleans and returns this list.
        Raises XyzPython.XyzMessage.DataError if the size in the buffer
        is not sufficient for the operation.
        '''
        if len(self.data) == 0:
            raise DataError("No data.")

        nbools = len(self.data)
        dtype = "{}?".format(nbools)
        return list(struct.unpack(dtype, self.data))

    def ints(self):
        '''
        Unpacks the data to a list of integers and returns this list.
        Raises XyzPython.XyzMessage.DataError if the size in the buffer
        is not sufficient for the operation.
        '''
        if len(self.data) == 0:
            raise DataError("No data.")
        if len(self.data)%4 != 0:
            raise DataError("Size of Data must be a multiple of 4.")

        nints = len(self.data) // 4
        dtype = "{}i".format(nints)
        return list(struct.unpack(dtype, self.data))
    
    def floats(self):
        '''
        Unpacks the data to a list of floats and returns this list.
        Raises XyzPython.XyzMessage.DataError if the size in the buffer
        is not sufficient for the operation.
        '''
        if len(self.data) == 0:
            raise DataError("No data.")
        if len(self.data)%4 != 0:
            raise DataError("Size of Data must be a multiple of 4.")

        nfloats = len(self.data) // 4
        dtype = "{}f".format(nfloats)
        return list(struct.unpack(dtype, self.data))

    def longs(self):
        '''
        Unpacks the data to a list of longs and returns this list.
        Raises XyzPython.XyzMessage.DataError if the size in the buffer
        is not sufficient for the operation.
        '''
        if len(self.data) == 0:
            raise DataError("No data.")
        if len(self.data)%8 != 0:
            raise DataError("Size of Data must be a multiple of 8.")

        nlongs = len(self.data) // 8
        dtype = "{}q".format(nlongs)
        return list(struct.unpack(dtype, self.data))

    def string(self):
        '''
        UTF-8 decodes the data to a string and returns this string.
        Raises XyzPython.XyzMessage.DataError if the size in the buffer
        is not sufficient for the operation.
        '''
        return self.data.decode("utf8")

    def strings(self):
        '''
        Splits the data at NULL-Bytes and returns UTF-8 decoded strings.
        Raises XyzPython.XyzMessage.DataError if the size in the buffer
        is not sufficient for the operation.
        '''
        return self.data.decode("utf8").split("\x00")

File: XyzJS [incomplete]/XyzMessage.js
const XyzUtils = require('./XyzUtils.js');

class XyzMessage {
    _bytes = null;
    _type = 0;

    /**
     * XyzMessage object
     * @param {Uint8Array} bytes
     * @param {Number} type
     */
    constructor(bytes, type=0) {
        this._bytes = bytes;
        this._type = type;
    }

    get bytes() {
        return this._bytes;
    }

    get type() {
        return Number(this._type);
    }
    
    get int() {
        return XyzUtils.Uint8ArrayToInt32(this._bytes);
    }

    get long() {
        return XyzUtils.Uint8ArrayToInt64(this._bytes);
    }

    get string() {
        return XyzUtils.TextDecoder.decode(this._bytes);
    }

    get strings() {
        return XyzUtils.TextDecoder.decode(this._bytes).split('\0');
    }

    get booleans() {
        const arr = [];

        for (let i=0; i<this._bytes; i++) {
            arr.push(this._bytes[i] == 1 ? true : false);
        }

        return arr;
    }

    get ints() {
        const arr = [];

        for (let i=0; i<this._bytes-4; i+=4) {
            const buffer = new Uint8Array(4);

            buffer.push(this._bytes[i]);
            buffer.push(this._bytes[i+1]);
            buffer.push(this._bytes[i+2]);
            buffer.push(this._bytes[i+3]);

            arr.push(XyzUtils.Uint8ArrayToInt32(buffer));
        }

        return arr;
    }

    get longs() {
        const arr = [];

        for (let i=0; i<this._bytes-8; i+=8) {
            const buffer = new Uint8Array(8);

            buffer.push(this._bytes[i]);
            buffer.push(this._bytes[i+1]);
            buffer.push(this._bytes[i+2]);
            buffer.push(this._bytes[i+3]);
            buffer.push(this._bytes[i+4]);
            buffer.push(this._bytes[i+5]);
            buffer.push(this._bytes[i+6]);
            buffer.push(this._bytes[i+7]);

            arr.push(XyzUtils.Uint8ArrayToInt64(buffer));
        }
        
        return arr;
    }
}

module.exports = XyzMessage;